cli/lsp/diagnostics.rs

```diff
@@ -25,8 +35,299 @@ pub enum DiagnosticSource {
+            .diagnostics_for(&specifier, &source)
+            .cloned()
+            .collect();
+          if tx.send(diagnostics).is_err() {
+            error!("DiagnosticServer unable to send response on channel.");
+          }
+        }
+        DiagnosticRequest::Invalidate(specifier) => {
+          collection.invalidate(&specifier)
+        }
+        DiagnosticRequest::Update(snapshot) => *maybe_snapshot = Some(snapshot),
+      }
+      true
+    }
+    _ => false,
+  }
+}
+
+/// A server which calculates diagnostics in its own thread and publishes them
+/// to an LSP client.
+#[derive(Debug)]
+pub struct DiagnosticsServer(mpsc::UnboundedSender<DiagnosticRequest>);
+
+impl DiagnosticsServer {
+  pub fn new(
+    client: Client,
+    ts_server: Arc<tsc::TsServer>,
+    performance: Performance,
+  ) -> Self {
+    let (tx, mut rx) = mpsc::unbounded_channel::<DiagnosticRequest>();
+    let _join_handle = thread::spawn(move || {
+      let runtime = create_basic_runtime();
+      let mut collection = DiagnosticCollection::default();
+
+      runtime.block_on(async {
+        // Some(snapshot) is the flag we use to determine if something has
+        // changed where we will wait for the timeout of changes or a request
+        // that forces us to update diagnostics
+        let mut maybe_snapshot: Option<StateSnapshot> = None;
+
+        loop {
+          let next = rx.recv();
+          tokio::pin!(next);
+
+          let duration = if maybe_snapshot.is_some() {
+            time::Duration::from_millis(DIAGNOSTIC_DEBOUNCE_MS)
+          } else {
+            // we need to await an arbitrary silly amount of time, so this is
+            // 1 year in seconds
+            time::Duration::new(31_622_400, 0)
+          };
+          let debounce = time::sleep(duration);
+          tokio::pin!(debounce);
+
+          // "race" the next message off the rx queue or the debounce timer.
+          // if the next message comes off the queue, the next iteration of the
+          // loop will reset the debounce future.  When the debounce future
+          // occurs, the diagnostics will be updated based on only the most
+          // recent snapshot, thereby "abandoning" any interim states.
+          tokio::select! {
+            _ = &mut debounce => {
+              if let Some(snapshot) = maybe_snapshot {
+                maybe_snapshot = None;
+                update_diagnostics(
+                  &client,
+                  &mut collection,
+                  &snapshot,
+                  &ts_server,
+                  &performance,
+                )
+                .await;
+              }
+              let maybe_request = next.await;
+              if !handle_request(maybe_request, &mut collection, &mut maybe_snapshot) {
+                break;
+              }
+            }
+            maybe_request = &mut next => {
+              if !handle_request(maybe_request, &mut collection, &mut maybe_snapshot) {
```

Comment:
I don't understand why handle_request is called here. 
Wouldn't you want to just loop around so that the debounce timer restarts?

Reply:
Because `next` isn't just update diagnostic requests, it is also "getting" and "updating" requests, so this processes those, and we need to break on a `None` return, because it means the channel has closed.